kong/plugins/rate-limiting/handler.lua

```diff
@@ -104,6 +111,18 @@ function RateLimitingHandler:access(conf)
         ngx.header[RATELIMIT_LIMIT .. "-" .. k] = v.limit
         ngx.header[RATELIMIT_REMAINING .. "-" .. k] = math.max(0, (stop == nil or stop == k) and v.remaining - 1 or v.remaining) -- -increment_value for this current request
       end
+
+      -- Find the closer limit.
+      local closer_limit = 0
+      for k, v in pairs(usage) do
+        if (closer_limit == 0 or closer_limit > v.remaining) then
+          closer_limit = v.remaining
+          ngx.header[RATELIMIT_LIMIT] = v.remaining
+          ngx.header[RATELIMIT_REMAINING] = v.limit
+          -- Add a little jitter to the expected reset time.
+          ngx.header[RATELIMIT_RESET] = math.random(0,10) + SECONDS_IN_PERIOD[k] - (current_timestamp/1000 % SECONDS_IN_PERIOD[k])
```

Comment:
A few notes here:
1. What is the reasoning behind the "jitter" inserted here? It seems unnecessary. The fixed window approach followed by this plugin should already guarantee that there is a given second at which the counters will be reset.
2. It is inefficient to call `ngx.header` within a loop. Instead, better to assign this value to a local, and assign `ngx.header` to this local _after_ the loop.
3. `SECONDS_IN_PERIODS` is missing `day`, `month`, and `year` values. This code will try to do arithmetic on a `nil` value and throw interpreter errors.

Reply:
My biggest concern with this PR relates to [a comment](https://github.com/Kong/kong/pull/2735/files#r134082010) I made last year in a similar PR (see the second part of the comment).

I believe we should show two headers to clients:
1. A `X-Rate-Limit-Retry`, which should only appear in `429` responses, and tells the client when will they be able to make a request again. This header should be inserted after having found the shortest limit that triggered a rejection (e.g. `second` over `minute`).
2. A `X-Rate-Limit-Reset`, which could be present in all responses sent to the client, and indicate when will _all_ of their counters be reset. In this case, this header would need to find the _largest_ limit configured (e.g. `minute` over `second`).

This way, a client has a way of knowing:
1. When is the soonest they can retry a request once they have been blocked
2. When can they expect their quota to be fully reset (i.e. when does the current time window ends?)

Here, is seems like you are trying to implement 1., but the chosen name and behavior contradicts those goals. Thoughts?

Reply:
* Yes to `Retry-After`, let's use this name if it is part of the RFC. I believe, as pointed out in my initial comment, that it should contain the value indicating to the client when, at the soonest, will they be able to make another request.

And as per the [RFC](https://tools.ietf.org/html/rfc6585#section-4), this seems check out:

>  The response representations SHOULD include details explaining the
   condition, and MAY include a Retry-After header indicating how long
   to wait before making a new request.

* `X-RateLimit-Reset`: imho, this should indicate to the client when _all_ of their counters will be reset (and thus provides the information `Retry-After` does not already).

* `X-RateLimit-Limit`: I still do not see the point of this header, and find it confusing because it fluctuates between time periods (minutes, seconds, etc...) depending on which one is the closest to `0`. I have trouble understanding how a client could make efficient use of such a header, since it cannot reasonably come up with an execution plan for future requests, because the time window of the counter is unspecified. Would you mind clarifying?
And again, my point about duplication of information still stands (this value is already present in the headers and can be deducted by clients).

Reply:
Ah, I think it just struck me that you wish `X-RateLimit-Limit` to be interpreted _alongside_ the `X-RateLimit-Reset` header, thus indicating to the client which time window is currently the most restrictive to them, and when such a time window will be reset... I am still not sure if this is free of confusion for instances of this plugin configured with different time windows... And my point about this information already being available to the clients still stands...

Reply:
@thibaultcha The rationale of adding the couple (Limit, Reset) to the "labeled" headers is here:

  - https://github.com/Kong/kong/pull/3451#issuecomment-388553141

Thx++, R

Reply:
The logic is the same, but honoring the largest period ff678caf5

HTH. Thanks again for all your time, R: